Task3/src/Repositories/GenericRepository/GenericRepositoryClass.hs

```diff
@@ -1,57 +1,96 @@
-{-# LANGUAGE AllowAmbiguousTypes #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
 {-# HLINT ignore "Redundant <&>" #-}
+{-# HLINT ignore "Redundant <$>" #-}
+{-# LANGUAGE TypeApplications #-}
+{-# HLINT ignore "Use >=>" #-}
 
 module Repositories.GenericRepository.GenericRepositoryClass (GenericRepository(..)) where
 
 import Utils.Files (readEntityFields, addLine, replaceLine, deleteLine)
-import Utils.Utils (maybeHead)
 import Data.List (findIndex)
-import Data.Maybe (fromMaybe)
 import Data.Converters.Converter
 import Data.RepositoryEntity.RepositoryEntity
 import Data.Functor ((<&>))
 import Data.SearchModels (SearchModel(..))
 import Repositories.FilterApplier (applyPagination)
+import Data.App (App, AppConfig (pageSize), AppState)
+import Control.Monad.Writer (MonadWriter(tell))
+import Control.Monad.Reader (MonadReader(ask))
+import qualified Control.Monad.State as S
+import Utils.Cache
+import Utils.Utils (validateArr, validateMaybe)
 
-class (ReadWriteEntity a, RepositoryEntity a) => GenericRepository a where
-    getList :: IO [a]
+class (ReadWriteEntity a, RepositoryEntity a, CacheAccess a) => GenericRepository a where
+    getList :: App [a]
     getList =
-        readEntityFields (entityString (entityName :: EntityName a)) >>=
-        mapM (return . readEntity)
+        tell ["getList: starting"] >>
+        S.get >>= \state ->
+        let cacheData = getCache state :: [a]
+        in getData (not (null cacheData)) state cacheData >>= \result ->
+        tell ["getList: completed"] >>
+        return result
+        where
+            getData :: Bool -> AppState -> [a] -> App [a]
+            getData True _ cache = return cache 
+            getData False state _ = 
+                (readEntityFields (entityString (entityName :: EntityName a)) >>= \fields ->
+                mapM (return . readEntity) fields) >>= \arr -> 
+                let newState = setCache state arr
+                in S.put newState >>
+                return arr 
 
-    get :: Int -> IO (Maybe a)
-    get eid = maybeHead . filter (\a -> entityId a == eid) <$> getList
+    get :: Int -> App a
+    get eid =
+        tell ["get: starting"] >>
+        filter (\a -> entityId a == eid) <$> getList >>= \entityArr ->
+        validateArr entityArr >>= \result -> 
+        tell ["get: completed"] >>
+        return result
 
-    add :: a -> IO Int
+    add :: a -> App Int
     add entity =
-        ((getList :: IO [a]) <&> getUnicId) >>= \entId ->
+        tell ["add: starting"] >>
```

Comment:
Let's add more descriptive logs everywhere, like "adding new customer" and "customer with id 4 has been added"

Reply:
And the same for other logs